### DISCUSSION TITLE
Training pipeline SentenceRecognizer

### DISCUSSION THREAD

**Discussion by JimDunlop** [explosion/spaCy#9110] [Help: Coding & Implementations] [answered]
I am trying to train the SentenceRecognizer. I generated the training data by using spacy sentence-segmentation on my dataset and writing that to a simple txt file.

    for row in df.itertuples():
        doc = nlp(row.Text)
        for sent in doc.sents:
            sentences.append(str(sent))
    with open('test_data.txt', 'w', encoding='utf-8') as f:
        for row in sentences:
            f.write(row+"\n")

Each "spacy-predicted-sentence" starts on a new line and I went through the file and corrected all missinterpretations. A typical Error would look like this:
Ich habe am Montag die Zähler-Nr. 
456 zur Warnung ausgeschrieben.

Correct Sentence: 
Ich habe am Montag die Zähler-Nr. 456 zur Warnung ausgeschrieben.

With the corrected data (Each line is one complete sentence now) i create the train.spacy file

    db = DocBin()
    sent_docs = [nlp(sent) for sent in sentences]
    doc = Doc.from_docs(sent_docs, ensure_whitespace=True)
    db.add(doc)
    db.to_disk("train.spacy")

I start spacy train:

    spacy train config.cfg --paths.train train.spacy --paths.dev=".\dev" -o output_dir

```
=========================== Initializing pipeline ===========================
[2021-09-01 12:37:53,867] [INFO] Set up nlp object from config
[2021-09-01 12:37:53,879] [INFO] Pipeline: ['senter']
[2021-09-01 12:37:53,883] [INFO] Created vocabulary
[2021-09-01 12:37:53,883] [INFO] Finished initializing nlp object
[2021-09-01 12:37:55,213] [INFO] Initialized pipeline components: ['senter']
✔ Initialized pipeline

============================= Training pipeline =============================
ℹ Pipeline: ['senter']
ℹ Initial learn rate: 0.001
E    #       LOSS SENTER  SENTS_F  SENTS_P  SENTS_R  SCORE
---  ------  -----------  -------  -------  -------  ------
c:\users\hua\miniconda3\lib\site-packages\spacy\training\corpus.py:77: UserWarning: [W090] Could not locate any .spacy files in path 'dev'.
  warnings.warn(Warnings.W090.format(path=orig_path, format=file_type))
  0       0      9633.00     0.00     0.00     0.00    0.00
200     200    274740.14     0.00     0.00     0.00    0.00
400     400      7935.46     0.00     0.00     0.00    0.00
600     600      6435.44     0.00     0.00     0.00    0.00
800     800      5806.15     0.00     0.00     0.00    0.00
1000    1000      5326.73     0.00     0.00     0.00    0.00
1200    1200      4976.42     0.00     0.00     0.00    0.00
1400    1400      4919.71     0.00     0.00     0.00    0.00
1600    1600      4721.10     0.00     0.00     0.00    0.00
✔ Saved pipeline to output directory
```

I load the model with

    nlp = spacy.load('trained_model\model-best')

and still get the same errors. I guess there is something fundamently wrong i am doing

**Comment by polm** [comment, marked as answer]
It's a little hard to tell what's going on here. 

```
 c:\users\hua\miniconda3\lib\site-packages\spacy\training\corpus.py:77: UserWarning: [W090] Could not locate any .spacy files in path 'dev'.
  warnings.warn(Warnings.W090.format(path=orig_path, format=file_type))
```

As this error indicates, it seems you don't have any valid dev data in the path you specified, which is why your evaluation output is all zero. If you don't have a dev set (you really should), you can specify your training data as the dev set to make sure some kind of learning is happening.

> With the corrected data (Each line is one complete sentence now) i create the train.spacy file

Is it correct that your input is now one sentence per Doc? That means the model could learn to never mark any token as the start of a sentence and be correct. You should stick a few sentences together to create multi-sentence docs so the model learns to split text, but only in the correct places.

However even with single-sentence docs I would expect to see some change in the output. Can you share your config? How many training examples do you have? 

Also note that based on your code here you are saving to `output_dir` but loading from `trained_model`. I assume that is a typo but if not that might explain the lack of change.

**Reply by polm** [reply]
Actually... since your evaluation scores are always zero, what may be happening is:

1. The initial model has a score of 0, which is "best", and is saved as `model-best`
2. It never gets better so it's not updated

Which means `model-best` is basically the same model you started with. 

You can fix that by fixing your evaluation data.

**Reply by JimDunlop** [reply]
Thanks again for your help. I got it working now. It was indeed the missing dev.spacy file. 
I changed my training data to include multiple sentences per line and copied the train.spacy to dev.spacy. The training output is now:
=========================== Initializing pipeline ===========================
[2021-09-02 10:28:37,169] [INFO] Set up nlp object from config
[2021-09-02 10:28:37,185] [INFO] Pipeline: ['senter']
[2021-09-02 10:28:37,189] [INFO] Created vocabulary
[2021-09-02 10:28:37,189] [INFO] Finished initializing nlp object
[2021-09-02 10:28:38,313] [INFO] Initialized pipeline components: ['senter']
✔ Initialized pipeline

============================= Training pipeline =============================
ℹ Pipeline: ['senter']
ℹ Initial learn rate: 0.001
E    #       LOSS SENTER  SENTS_F  SENTS_P  SENTS_R  SCORE
---  ------  -----------  -------  -------  -------  ------
  0       0        44.50     1.13     0.70     3.07    0.01
  0     200      1705.83    87.79    87.72    87.85    0.88
  0     400       364.32    92.19    90.66    93.77    0.92
  1     600       315.90    95.61    95.93    95.30    0.96
  1     800       242.62    96.20    95.70    96.70    0.96
  2    1000       244.00    97.87    97.72    98.02    0.98
  3    1200       201.98    98.35    98.47    98.22    0.98
  4    1400       216.20    98.64    98.41    98.88    0.99
  6    1600       192.77    98.93    98.86    99.01    0.99
  7    1800       223.03    99.10    99.21    98.98    0.99
 10    2000       229.81    99.01    99.11    98.91    0.99
 12    2200       261.57    99.25    99.16    99.34    0.99
 15    2400       252.80    99.33    99.29    99.36    0.99
 19    2600       248.33    99.36    99.36    99.36    0.99
 22    2800       194.51    99.36    99.31    99.42    0.99
 25    3000       222.05    99.33    99.29    99.36    0.99
 29    3200       220.95    99.36    99.31    99.42    0.99
 32    3400       206.42    99.37    99.24    99.49    0.99
 36    3600       207.10    99.17    99.26    99.09    0.99
 39    3800       192.47    99.44    99.34    99.54    0.99
 42    4000       180.24    99.36    99.31    99.42    0.99
 46    4200       177.36    99.37    99.24    99.49    0.99
 49    4400       166.50    99.33    99.36    99.29    0.99
 52    4600       160.85    99.44    99.42    99.47    0.99
 56    4800       158.34    99.44    99.42    99.47    0.99
 59    5000       154.84    99.44    99.42    99.47    0.99
 62    5200       145.70    99.44    99.42    99.47    0.99
 66    5400       154.77    99.44    99.34    99.54    0.99
✔ Saved pipeline to output directory

A first test with the new pipeline looked promising.

**Reply by JimDunlop** [reply]
If i understand it correctly, i have created a new model, which only includes the senter as a pipeline component and a vocab (from the training data i guess).
So if wanted to use my updated senter-component in another pipeline it try:

    text = "Ich habe am Montag die Zähler-Nr. 456 zur Warnung ausgeschrieben."
    nlp = spacy.load('de_core_news_sm', exclude=['senter'])
    senter_source = spacy.load('output_dir/model-best')
    nlp.add_pipe('senter',source=senter_source)
    nlp.enable_pipe('senter')
    doc = nlp(text)

    for sent in doc.sents:
        print(sent)

Unfortunatly it gives me the same wrong sentence splitting.
Ich habe am Montag die Zähler-Nr.
456 zur Warnung ausgeschrieben.

If i load the senter directly with nlp = spacy.load it splits the sentences corrctly. So either the component does not get overritten, or i am missing something.

**Reply by polm** [reply]
Change the `add_pipe` line to be like this:

    nlp.add_pipe('senter',source=senter_source, before="parser")

By default components are added at the end of the pipeline. But another rule of components is that (by default) they respect any data that was set before them - since the parser sets sentence boundaries, if the senter is after it it won't do anything. 

Also, don't use `exclude=["senter"]` - that doesn't do anything because the default pipeline has no senter. Similarly `enable_pipe` does nothing so you should also remove that.

It's a pretty different use case, but maybe look at [this example project](https://github.com/explosion/projects/tree/v3/tutorials/ner_double#notes-on-combining-components) for an example of combining components from different pipelines - it touches on some of these points, and there's example code and an equivalent config.

**Reply by JimDunlop** [reply]
before='parser' works like a charm! Thanks
without the exclude parameter i get the error

    raise ValueError(Errors.E007.format(name=name, opts=self.component_names))
ValueError: [E007] 'senter' already exists in pipeline. Existing names: ['tok2vec', 'tagger', 'morphologizer', 'parser', 'senter', 'attribute_ruler', 'lemmatizer', 'ner']

**Reply by polm** [reply]
Huh, that's weird about exclude. I just installed the v3.1.0 version of `de_core_news_sm` and it has the same components, except it has no senter. Maybe you installed a customized version of the pipeline, or trained a model to a directory with that name...?

**Reply by JimDunlop** [reply]
I will except it as it is. :)
If i run into probems with a future pipeline i know what to do. Again, thanks a lot
